source/workflowapi/app.py

```diff
@@ -2230,6 +2414,505 @@ def update_workflow_execution_status(id, status, message):
+    return response
+
+# ===== ACT
+@app.route('/service/translate/get_parallel_data', cors=True, methods=['POST'], content_types=['application/json'], authorizer=authorizer)
+def get_parallel_data():
+    """ Get a link to the CSV formatted description for an Amazon Translate Parallel Data Set.
+
+    Body:
+
+    .. code-block:: python
+
+        {
+            'parallel_data_name'='string'
+        }
+
+    Returns:
+        This is a proxy for boto3 get_parallel_data and returns the output from that SDK method.
+        See `the boto3 documentation for details <https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/translate.html#Translate.Client.get_parallel_data>`_
+
+    Raises:
+        See the boto3 documentation for details
+        500: ChaliceViewError - internal server error
+    """
+    print('get_parallel_data request: '+app.current_request.raw_body.decode())
+    translate_client = boto3.client('translate', region_name=os.environ['AWS_REGION'])
+    parallel_data_name = json.loads(app.current_request.raw_body.decode())['parallel_data_name']
+    response = translate_client.get_parallel_data(Name=parallel_data_name, parallel_dataDataFormat='CSV')
+    # Remove response metadata since we don't need it
+    if 'RespnseMetadata' in response:
+        del response['ResponseMetadata']
+    # Convert time field to a format that is JSON serializable
+    response['TerminologyProperties']['CreatedAt'] = response['TerminologyProperties']['CreatedAt'].isoformat()
+    response['TerminologyProperties']['LastUpdatedAt'] = response['TerminologyProperties']['LastUpdatedAt'].isoformat()
+    return response
+
+
+@app.route('/service/translate/download_parallel_data', cors=True, methods=['POST'], content_types=['application/json'], authorizer=authorizer)
+def download_parallel_data():
+    """ Get the CSV formated contents of an Amazon Translate Parallel Data Set.
+
+    Body:
+
+    .. code-block:: python
+
+        {
+            'parallel_data_name'='string'
+        }
+
+
+    Returns:
+        A string contining the CSV formatted Amazon Transcribe parallel_data
+
+        .. code-block:: python
+
+            {
+                'parallel_data_csv': string
+            }
+
+    Raises:
+        See the boto3 documentation for details
+        500: ChaliceViewError - internal server error
+    """
+    # This function returns the specified parallel_data in CSV format, wrapped in a JSON formatted response.
+    print('download_parallel_data request: '+app.current_request.raw_body.decode())
+    translate_client = boto3.client('translate', region_name=os.environ['AWS_REGION'])
+    parallel_data_name = json.loads(app.current_request.raw_body.decode())['parallel_data_name']
+    url = translate_client.get_parallel_data(Name=parallel_data_name, ParallelDataFormat='CSV')['ParallelDataLocation']['Location']
+    import urllib.request
+    parallel_data_csv = urllib.request.urlopen(url).read().decode("utf-8")
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *urlopen*. The resource is allocated by call *request.urlopen*. Execution paths that do not contain closure statements were detected. To prevent this resource leak, close the object returned by *urlopen()* in a try-finally block or declare it using a `with` statement.

**More info**
[View details about the `with` statement in the Python developer&#39;s guide](https://www.python.org/dev/peps/pep-0343/) (external link).

Reply:
We are not going to address this CodeGuru finding because we want to mirror the official AWS code sample for sigv4 signing: https://docs.aws.amazon.com/code-samples/latest/catalog/python-signv4-v4-signing-get-post.py.html.